### PR TITLE
RUM-7549 Handle trace with coroutines

### PR DESCRIPTION
**Pull request by xgouchet**
### What does this PR do?

The main focus is to avoid sharing the same list of scopes across tracers, and add some internal logs to better diagnose issues with tracing.

### CODE REVIEW

File: features/dd-sdk-android-trace/src/main/java/com/datadog/opentracing/PendingTrace.java [hunk 1]
```diff
@@ -84,9 +89,27 @@ public long getCurrentTimeNano() {
...
+        false,
+        new HashMap<>()
+      );
       return;
     }
     if (!traceId.equals(span.context().getTraceId())) {
+      internalLogger.log(
+        InternalLogger.Level.ERROR,
+        InternalLogger.Target.USER,
+        () -> "Span " + span.getOperationName() + " not registered because of traceId mismatch; " +
```
Comment (0xnm) [inline features/dd-sdk-android-trace/src/main/java/com/datadog/opentracing/PendingTrace.java:107]: is it useful here also to log the value of `span.context().getTraceId()`?

File: features/dd-sdk-android-trace/src/main/java/com/datadog/opentracing/PendingTrace.java [hunk 2]
```diff
@@ -96,19 +119,55 @@ public void registerSpan(final DDSpan span) {
...
+          false,
+          new HashMap<>()
+      );
       return;
     }
     if (!traceId.equals(span.context().getTraceId())) {
+      internalLogger.log(
+          InternalLogger.Level.ERROR,
+          InternalLogger.Target.USER,
+          () -> "Span " + span.getOperationName() + " not expired because of traceId mismatch; " +
```
Comment (0xnm) [inline features/dd-sdk-android-trace/src/main/java/com/datadog/opentracing/PendingTrace.java:152]: same as above

File: features/dd-sdk-android-trace/src/main/java/com/datadog/opentracing/PendingTrace.java [hunk 3]
```diff
@@ -84,9 +89,27 @@ public long getCurrentTimeNano() {
 
   public void registerSpan(final DDSpan span) {
     if (traceId == null || span.context() == null) {
+      internalLogger.log(
+        InternalLogger.Level.ERROR,
```
Comment (0xnm) [inline features/dd-sdk-android-trace/src/main/java/com/datadog/opentracing/PendingTrace.java:93]: In `DDSpan` when we are dropping the span `warn` level is used, but here it is an `error` in the whole file. Should they be aligned?

**Review by 0xnm — APPROVED** [review]
I see that these changes only for OpenTracing, should we do something similar for OpenTelemetry.

Anyway, lgtm.

File: features/dd-sdk-android-trace/src/main/java/com/datadog/opentracing/PendingTrace.java (hunk 3 above)
Comment (xgouchet) [reply features/dd-sdk-android-trace/src/main/java/com/datadog/opentracing/PendingTrace.java:93]: In DDSpan, we use Warning as the cause of the drop is (most likely) misuse of the SDK, usually by calling `finish()` more than once.
Here the drop reason are ones that _should not happen™️_, hence the use of Error level.

File: features/dd-sdk-android-trace/src/main/java/com/datadog/opentracing/PendingTrace.java (hunk 1 above)
Comment (xgouchet) [reply features/dd-sdk-android-trace/src/main/java/com/datadog/opentracing/PendingTrace.java:107]: Makes sense, yes.

File: features/dd-sdk-android-trace/src/main/java/com/datadog/opentracing/PendingTrace.java (hunk 2 above)
Comment (xgouchet) [reply features/dd-sdk-android-trace/src/main/java/com/datadog/opentracing/PendingTrace.java:152]: 👍

**Pull request merged**